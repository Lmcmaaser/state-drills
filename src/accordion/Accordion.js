import React, { Component } from 'react'
import './Accordion.css'

export default class Accordian extends Component {
    static defaultProps = {  //defaultProps: when a prop is missing it displays the fallback value in place of the missing prop
        sections: []  //sections is the prop, and an array of objects
    };

    state = {
        stateSectionIndex: null,
    }

    handleStateSetSection = (sectionIndex) => {
        this.setState({stateSectionIndex: sectionIndex}) //'this' references a JavaScript element
    }

    renderItem(section, stateSectionIndex, idx) {  //idx is a utility function for traversing properties on objects and arrays.
        return (
            <li className='accordionItem' key={idx}>
                <button
                    type='button'
                    onClick={() => this.handleStateSetSection(idx)}
                >
                    {section.title}
                </button>
                    {(stateSectionIndex === idx) && <p>{section.content}</p>}
            </li>
        )
    }

    render() {
        const { stateSectionIndex } = this.state
        const { sections } = this.props 
        return (
            <ul className='Accordion'>
                {sections.map((section, idx) => 
                    this.renderItem(section, idx, stateSectionIndex)
                )}
            </ul>
        )
    }
};