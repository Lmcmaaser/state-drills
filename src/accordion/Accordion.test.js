import React from 'react'
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme'  //shallow function in Enzyme will create a wrapper instance of our component that we can interact with
import toJson from 'enzyme-to-json'
import Accordion from './Accordion'

const sectionsProp = [
  {
    title: 'Section 1',
    content: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.',
  },
  {
    title: 'Section 2',
    content: 'Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!',
  },
  {
    title: 'Section 3',
    content: 'Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?',
  },
]

describe(`Accordion component`, () => {
    //smoke test, ensures that the component renders in the first place
    it('renders without crashing', () => {
        // first create a DOM element to render the component into
        const ul = document.createElement('ul');
        //render the component, this is the actual test, if something is wrong it will fail here
        ReactDOM.render(<Accordion />, ul);
        //clean up code
        ReactDOM.unmountComponentAtNode(ul);
    })
    //snapshot test
    it('renders empty given no props without errors', () => {
        const wrapper = shallow(<Accordion />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('renders the no sections as default', () => {
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    it('opens clicked section', () => {
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        wrapper.find('button').at(1).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    it('opens one section at a time', () => {
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        wrapper.find('button').at(1).simulate('click')
        wrapper.find('button').at(2).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })
})