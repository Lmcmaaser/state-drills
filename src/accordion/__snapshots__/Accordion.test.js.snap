// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component opens clicked section 1`] = `
<ul
  className="Accordion"
>
  <li
    className="accordionItem"
    key="null"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 1
    </button>
  </li>
  <li
    className="accordionItem"
    key="null"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 2
    </button>
  </li>
  <li
    className="accordionItem"
    key="null"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion component opens one section at a time 1`] = `
<ul
  className="Accordion"
>
  <li
    className="accordionItem"
    key="null"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 1
    </button>
  </li>
  <li
    className="accordionItem"
    key="null"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 2
    </button>
  </li>
  <li
    className="accordionItem"
    key="null"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion component renders empty given no props without errors 1`] = `
<ul
  className="Accordion"
/>
`;

exports[`Accordion component renders the no sections as default 1`] = `
<ul
  className="Accordion"
>
  <li
    className="accordionItem"
    key="null"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 1
    </button>
  </li>
  <li
    className="accordionItem"
    key="null"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 2
    </button>
  </li>
  <li
    className="accordionItem"
    key="null"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 3
    </button>
  </li>
</ul>
`;
